src/Factory/SubjectUrl/ProductUrlGenerator.php

```diff
@@ -19,13 +19,17 @@ public function __construct(RouterInterface $router)
         $this->router = $router;
     }
 
-    public function can(RichSnippetSubjectInterface $subject): bool
+    public function can($subject): bool
     {
         return $subject instanceof ProductInterface;
     }
 
-    public function generateUrl(RichSnippetSubjectInterface $subject): string
+    public function generateUrl($subject): string
     {
+        if (!$this->can($subject)) {
+            throw new \LogicException('This case only works with a subject which implement ProductInterface');
+        }
+
         return $this->router->generate('sylius_shop_product_show', ['slug' => $subject->getSlug()], UrlGeneratorInterface::ABSOLUTE_URL);
```

Comment:
On a toujours ce `$subjcet->getSlug()` qui ne respect aucun contrat :/

Reply:
Avant, il avait un contrat qu'il ne respectait pas. Maintenant, il n'y a plus de contrat. Le garde fou est le $this->can